### DISCUSSION TITLE
Global Parameters in ARMI

### DISCUSSION THREAD

**Discussion by opotowsky** [terrapower/armi#1360] [General]
Global parameters cause issues. Mostly we run into said issues during unit testing, and global params prevent or run into issues with parallel unit test execution. Also, it is [against the standards and practices of ARMI]([‎doc/developer/standards_and_practices.rst](https://github.com/terrapower/armi/blob/f947ea12266b688f3f2e3402fa458076849d1375/doc/developer/standards_and_practices.rst#L49)), so maybe it's high time we follow our own standards! We need to keep some of these, but let's review them all and see if there's anything we can improve. (I'm lookin at you, caseSettings and nuclideBases!)

1. Nuclides

- [ ] global elements [‎armi/nucDirectory/elements.py](https://github.com/terrapower/armi/blob/f947ea12266b688f3f2e3402fa458076849d1375/armi/nucDirectory/elements.py#L398)
- [ ] many global nuclide params [armi/nucDirectory/nuclideBases.py](https://github.com/terrapower/armi/blob/f947ea12266b688f3f2e3402fa458076849d1375/armi/nucDirectory/nuclideBases.py#L1279)
- [ ] !!! [‎armi/nucDirectory/thermalScattering.py](https://github.com/terrapower/armi/blob/f947ea12266b688f3f2e3402fa458076849d1375/armi/nucDirectory/thermalScattering.py#L222) This one actually doesn't get deleted during the above destroy function!!!!!

2. Case settings & Blueprints

- [x] global Settings.instance [‎armi/settings/caseSettings.py](https://github.com/terrapower/armi/blob/f947ea12266b688f3f2e3402fa458076849d1375/armi/settings/caseSettings.py#L69) ([PR](https://github.com/terrapower/armi/pull/1399))
- [x] global AssemNum [armi/reactor/assemblies.py](https://github.com/terrapower/armi/blob/f947ea12266b688f3f2e3402fa458076849d1375/armi/reactor/assemblies.py#L41) ([PR](https://github.com/terrapower/armi/pull/1383))

3. Misc

- [x] global white space [‎armi/runLog.py](https://github.com/terrapower/armi/blob/f947ea12266b688f3f2e3402fa458076849d1375/armi/runLog.py#L127) ([PR](https://github.com/terrapower/armi/pull/1370))
- [x] global plugin flags registered [armi/reactor/flags.py](https://github.com/terrapower/armi/blob/f947ea12266b688f3f2e3402fa458076849d1375/armi/reactor/flags.py#L312) ([PR](https://github.com/terrapower/armi/pull/1388))
- [ ] global serial num [armi/reactor/parameters/parameterCollections.py](https://github.com/terrapower/armi/blob/f947ea12266b688f3f2e3402fa458076849d1375/armi/reactor/parameters/parameterCollections.py#L168)
- [ ] global test reactor [‎armi/reactor/tests/test_reactors.py](https://github.com/terrapower/armi/blob/f947ea12266b688f3f2e3402fa458076849d1375/armi/reactor/tests/test_reactors.py#L171) _(Hm, maybe this is the issue and not the case settings? haven't looked into this in awhile)_

4. Prob want to keep?

- [ ] mpiActions -- but that's a whole 'nother thing.
- [ ] 4 global params [‎armi/context.py](https://github.com/terrapower/armi/blob/f947ea12266b688f3f2e3402fa458076849d1375/armi/context.py#L84) (and line 180)
- [ ] material namespace order [‎armi/materials/__init__.py](https://github.com/terrapower/armi/blob/f947ea12266b688f3f2e3402fa458076849d1375/armi/materials/__init__.py#L50)

**Comment by john-science** [comment]
I totally agree. All (or maybe nearly all, we'll see) of these were just someone trying to save 10 minutes a few years ago, and now we're stuck with it.

Some of these are only bad ideas (looking at you `_WHITE_SPACE` in `runLog.py`).  But some of these make running parallel ARMI cases significantly harder, or make having two ARMI reactors in memory at the same time impossible (looking at you case settings, blueprints, and plugins).

**Reply by john-science** [reply]
@drewj-usnctech Just FYI.

**Reply by drewj-usnctech** [reply]
Generally okay removing globals. And if it breaks something, we probably shouldn't have done it in the first place

**Comment by john-science** [comment]
Could we move `parameterCollections.GLOBAL_SERIAL_NUM` onto the `Reactor` or something?

**Reply by opotowsky** [reply]
![image](https://github.com/terrapower/armi/assets/5347433/8e828242-7312-4c6e-82d8-57cc5889aab4)

Looks like it's only used in the database, that keeps track of its own `layout.serialNum`

**Reply by john-science** [reply]
Yeah, something like that:

https://github.com/terrapower/armi/blob/e8ddcdf774f9eb3e842ae75948f095095fc2d48b/armi/bookkeeping/db/database3.py#L1107
https://github.com/terrapower/armi/blob/e8ddcdf774f9eb3e842ae75948f095095fc2d48b/armi/bookkeeping/db/database3.py#L1276-L1277

I just wonder if we can pull it out of the global Python space and put it into the Reactor (or maybe DB?) object somewhere.  Just something so it is not global space.  That way we could use this in MPI cases and we could allow for multiple reactors in memory at the same time.

**Reply by john-science** [reply]
Okay, I think it's clear the best option is to move the `assemblies.py:_assemNum` onto the `Reactor` object:

https://github.com/terrapower/armi/blob/f947ea12266b688f3f2e3402fa458076849d1375/armi/reactor/assemblies.py#L36-L44

And we can do the same for the `parameterCollection.GLOBAL_SERIAL_NUM`:

https://github.com/terrapower/armi/blob/e8ddcdf774f9eb3e842ae75948f095095fc2d48b/armi/reactor/parameters/parameterCollections.py#L34

The only question is:

> Should these be attributes of the `Reactor` class, or `Parameters`?

Making them each a `Parameter` is slightly more work.  But it would also mean they were stored in the database, which is probably nice/useful.

**Reply by opotowsky** [reply]
I would need someone else's input on if it would be a useful param. My knee-jerk reaction is that if it's not in the DB now, it doesn't need to be. In that case, I vote you make this a reactor attribute!

**Comment by albeanth** [comment]
> many global nuclide params [armi/nucDirectory/nuclideBases.py](https://github.com/terrapower/armi/blob/f947ea12266b688f3f2e3402fa458076849d1375/armi/nucDirectory/nuclideBases.py#L1279)

This one reared its head for @wcscherer. Just bringing this convo back to life to maybe knock some of these out in the new year.

**Reply by opotowsky** [reply]
Adding a vote here to prioritize the global nuclides removal, as it has significant downstream impacts and prevents our unit tests from being parallelized.

**Reply by john-science** [reply]
@wcscherer @albeanth @opotowsky 

I think that would be a solid improvement to the nucDir system.  Help me build manager interest in this.

**Reply by jakehader** [reply]
@john-science - I'm in agreement of removing globals and moving in a direction where we can serialize any particular state of the system to the database, including enhancing the parameter system.

**Reply by john-science** [reply]
@jakehader I will try to convince management of WHY this is important, so I can put this work on the calendar for ARMI version `0.4.0`.
